src/Bicep.Core/PrettyPrint/DocumentBuildVisitor.cs

```diff
@@ -0,0 +1,366 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parser;
+using Bicep.Core.PrettyPrint.Documents;
+using Bicep.Core.PrettyPrint.DocumentCombinators;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.PrettyPrint
+{
+    public class DocumentBuildVisitor : SyntaxVisitor
+    {
+        private static readonly ILinkedDocument Nil = new NilDocument();
+
+        private static readonly ILinkedDocument Space = new TextDocument(" ", Nil);
+
+        private static readonly ILinkedDocument Line = new NestDocument(0, Nil);
+
+        private static readonly IReadOnlyDictionary<string, TextDocument> CommonTextCache;

```

Comment:
>IReadOnlyDictionary [](start = 32, length = 19)

nit - should use `ImmutableDictionary`

Reply:
Changed it to use `ImmutableDictionary`.